Add support for grouping media players

AlexGustafsson opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue or discussion
  in the additional information section.
-->

Add support to the media player card and dialogs to group media players
that support grouping. Allows for SONOS speakers to be managed without
using custom components, services or automations.

![image](https://github.com/user-attachments/assets/b72b8b2b-8a56-4aaf-8cb0-6ceb1b52cc3c)

![image](https://github.com/user-attachments/assets/0140795d-3009-4c1a-beb7-d229bd39d9a6)

![image](https://github.com/user-attachments/assets/1a0138bb-bd4e-415c-b7f9-e86b462878d2)

## Type of change
<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

N/A.

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

None.

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

AlexGustafsson committed:
Add support for grouping media players

Add support to the media player card and dialogs to group media players
that support grouping. Allows for SONOS speakers to be managed without
using custom components, services or automations.

Co-authored-by: Simon Lamon <32477463+silamon@users.noreply.github.com>

AlexGustafsson commented:
Being my first contribution to the Home Assistant frontend, I might be missing some things that are obvious to others. I'm very grateful for any help with understanding how to go about getting this feature done in a good way.
 
 At this point, the feature is "complete", but lacks in polish and likely other things that I'm not aware of. My hope is to get some pointers on what needs to be done and if, at all, this feature is something the maintainers would want to merge.
 
 Some questions / observations from my end:
 
 - Setting `hide-clear-icon` on a `ha-entity-picker` does nothing. The attribute seems to be defined, but not implemented. So the "this speaker entity picker" can be cleared, which is a bit confusing.
 - Some UI elements seem to be misplaced (see screenshot of the selection, `ha-entities-picker`), the arrows of the combo boxes are a bit wonky. Am I missing some CSS?
 - Are there tests that should be written?

silamon commented:
Thank you for the contribution. At this point, the Needs UX has been added to gather feedback if the experience can be changed but from a feature it actually works like it should. 

> Some questions / observations from my end:
> 
> * Setting `hide-clear-icon` on a `ha-entity-picker` does nothing. The attribute seems to be defined, but not implemented. So the "this speaker entity picker" can be cleared, which is a bit confusing.
> * Some UI elements seem to be misplaced (see screenshot of the selection, `ha-entities-picker`), the arrows of the combo boxes are a bit wonky. Am I missing some CSS?
> * Are there tests that should be written?

The first observation seems to be right, the attribute no longer hides the clear button. 
I don't have any CSS issues, the display looks fine for me. 
We don't have much unit tests, so it's not a requirement to get it merged.

silamon commented on src/components/media-player/dialog-media-player-group.ts:
```diff
@@ -0,0 +1,155 @@
+        @opened=${this._dialogOpened}
+      >
+        <ha-dialog-header show-border slot="heading">
+          <span slot="title"> ${name} </span>
```
We should give it another name than the name of the media player. 
The title should be "Group" I suppose or "Group members" although I'm doubt about it.

silamon commented on src/components/media-player/dialog-media-player-group.ts:
```diff
@@ -0,0 +1,155 @@
+          .includeDomains=${includeDomains}
+          .entityFilter=${this._filterEntities}
+          .pickedEntityLabel=${this.hass.localize(
+            "ui.panel.config.person.detail.device_tracker_picked"
```
This should be a new label. You can make it under "ui.card.media_player.group"

silamon commented on src/components/media-player/dialog-media-player-group.ts:
```diff
@@ -0,0 +1,155 @@
+            "ui.panel.config.person.detail.device_tracker_picked"
+          )}
+          .pickEntityLabel=${this.hass.localize(
+            "ui.panel.config.person.detail.device_tracker_pick"
```
Same for this label.

silamon commented on src/components/media-player/dialog-media-player-group.ts:
```diff
@@ -0,0 +1,155 @@
+        <mwc-button slot="secondaryAction" @click=${this.closeDialog}>
+          ${this.hass.localize("ui.common.close")}
+        </mwc-button>
+        <mwc-button
+          .disabled=${this._submitting}
+          slot="primaryAction"
+          @click=${this._submit}
+        >
+          ${this.hass.localize("ui.common.save")}
+        </mwc-button>
```
Use the ha-button, so we import the css definitions of the home assistant buttons. 

```suggestion
        <ha-button
          .disabled=${this._submitting}
          slot="primaryAction"
          @click=${this._submit}
        >
          ${this.hass.localize("ui.common.save")}
        </ha-button>
```

Add an import for ha-button like: 
```
import "../ha-button";
```

silamon commented on src/components/media-player/dialog-media-player-group.ts:
```diff
@@ -0,0 +1,155 @@
+          @value-changed=${this._groupMembersChanged}
+        >
+        </ha-entities-picker>
+        <mwc-button slot="secondaryAction" @click=${this.closeDialog}>
+          ${this.hass.localize("ui.common.close")}
+        </mwc-button>
```
```suggestion
```
There are multiple close buttons in the dialog. One is enough :)

silamon commented on src/components/media-player/dialog-media-player-group.ts:
```diff
@@ -0,0 +1,155 @@
+  }
+
+  static get styles(): CSSResultGroup {
+    return [haStyleDialog, css``];
```
```suggestion
    return [haStyleDialog];
```

silamon commented on src/dialogs/more-info/controls/more-info-media_player.ts:
```diff
@@ -81,6 +83,20 @@ class MoreInfoMediaPlayer extends LitElement {
+        ${!isUnavailableState(stateObj.state) &&
+        supportsFeature(stateObj, MediaPlayerEntityFeature.GROUPING)
+          ? html`
+              <mwc-button
+                .label=${this.hass.localize("ui.card.media_player.group")}
+                @click=${this._showGroupMediaPlayers}
+              >
+                <ha-svg-icon
+                  .path=${mdiSpeakerMultiple}
+                  slot="icon"
+                ></ha-svg-icon>
+              </mwc-button>
```
```suggestion
              <ha-button
                .label=${this.hass.localize("ui.card.media_player.group")}
                @click=${this._showGroupMediaPlayers}
              >
                <ha-svg-icon
                  .path=${mdiSpeakerMultiple}
                  slot="icon"
                ></ha-svg-icon>
              </ha-button>
```

silamon replied:
This one is unresolved.

AlexGustafsson replied:
The button above that one (the browse media one) is an `mwc-button`. Should I

1. Change just the addition as suggested
2. Change both (the other one in a separate commit?)
3. Leave as is

silamon replied:
You can change both if you want, but sooner or later we're going to change it anyway to a ha-button

AlexGustafsson commented:
@silamon thanks for taking a look and guiding me! I've implemented the fixes. I went with "Group member" for the entity pickers and "Group players" for the card title. I've updated the applicable screenshot.

How do you typically use "resolve" in PRs? Do you want to use it to check things off when reviewing again or for contributors to mark that they fixed the comment?

AlexGustafsson commented:
I've rebased this PR as #24821 was merged. I've updated the screenshot as well. Now you can no longer clear the group leader entity picker.

silamon commented on src/translations/en.json:
```diff
@@ -216,7 +216,10 @@
-        "nothing_playing": "Nothing playing"
+        "nothing_playing": "Nothing playing",
+        "group": "Group",
+        "group_players": "Group players",
```
```suggestion
        "group_players": "Group media players",
```

AlexGustafsson replied:
Should we clarify the key as well, or is it fine to just have `group_players` as it's in the `media_player` object?

silamon replied:
You can change it if you want since it's a new key, otherwise we prefer to keep it since the old translation is then kept for the translators to iterate on.

AlexGustafsson committed:
Use ha-button in media player more info dialog

AlexGustafsson commented:
Rebased on dev again.

silamon commented:
> Rebased on dev again.

I've been poking for the ux review but they are still discussing.

bramkragten commented on src/components/media-player/dialog-media-player-group.ts:
```diff
@@ -0,0 +1,158 @@
+  }
+
+  private _dialogOpened() {
+    this.classList.add("opened");
```
Why do we add this class?

bramkragten commented on src/components/media-player/dialog-media-player-group.ts:
```diff
@@ -0,0 +1,158 @@
+    this._entityId = undefined;
+    this._groupMembers = [];
+    this._submitting = false;
+    this.classList.remove("opened");
```
And remove it

bramkragten commented on src/components/media-player/dialog-media-player-group.ts:
```diff
@@ -0,0 +1,158 @@
+    this._entityId = params.entityId;
+
+    const stateObj = this.hass.states[params.entityId] as
+      | (HassEntity & { attributes: { group_members?: string[] } })
```
```suggestion
      | MediaPlayerEntity
```

Update the `MediaPlayerEntity` (`MediaPlayerEntityAttributes`) interface with the `group_members` instead.

bramkragten commented on src/components/media-player/dialog-media-player-group.ts:
```diff
@@ -0,0 +1,158 @@
+      await mediaPlayerJoin(this.hass, this._entityId, this._groupMembers);
+      this.closeDialog();
+    } catch (_e) {
+      // TODO
```
show the error in an `ha-alert`

bramkragten commented:
Thanks for this! Great feature!

We should find a way to show that the entity is grouped with others atm, maybe with a different color of the button, or a badge on the button with the number of entities it is grouped with?

AlexGustafsson commented:
Here's some quick examples of the different choices.

Badge:

![image](https://github.com/user-attachments/assets/a19b4e68-1164-4ece-bbac-040282a480e9)

Color:
![image](https://github.com/user-attachments/assets/67fe73a3-af36-47e9-bba9-6062c5b6639c)

Label:
![image](https://github.com/user-attachments/assets/5295aad3-1758-42bb-b9f0-3affb2f927d8)

I'm thinking the last option is the clearest. It would also work to have "grouped" variants of the state icons, but those are not provided in the material design icon library. Having a badge for the button seems to require a bit of refactoring. I can't find anywhere else a badge is displayed close to a button's icon or have a color that would work with the white icons, they typically seem to have quite some margin?

bramkragten commented:
The badge is used in the sidebar for the settings and notification icons.

![CleanShot 2025-04-26 at 22 30 52@2x](https://github.com/user-attachments/assets/373b9c2b-5840-44ed-8e01-cc7f37a2a35e)

I kinda like the 3rd option too!

bramkragten commented on src/components/media-player/dialog-media-player-group.ts:
```diff
@@ -80,6 +85,9 @@ class DialogMediaPlayerGroup extends LitElement {
           ></ha-icon-button>
         </ha-dialog-header>
+        ${this._error
+          ? html`<ha-alert alert-type="error">${this._error}</ha-alert>`
```
You also import `ha-alert`

AlexGustafsson replied:
Thanks, I've added it now. I tested it when adding it and it seems to work without. Why is that? The component just happens to be imported by something else in the tree or?

silamon replied:
You should add the component import always. That may be possible if you've visited a different page that loads the same ha-alert component, then it shows up. If that's not the case, you won't be able to see the ha-alert, since it doesn't know yet where the component is defined.

AlexGustafsson commented:
> The badge is used in the sidebar for the settings and notification icons.

Those seem to be using list items, which in turn use `@material/web` components which have a slot for the icon. Other buttons don't seem to have that. I've tried to replicate the looks. Take a look in the code and see if it's alright. It's basically the same badge class as used everywhere else, but with absolute position as the button component doesn't have a similar slot.

<img width="511" alt="image" src="https://github.com/user-attachments/assets/2a8856e5-8d3b-47a2-a22f-333e24152205" />

Given that groups represent a "normal" state, perhaps at least the color should change?

matthiasdebaat commented:
Thanks for this PR @AlexGustafsson! We have a few suggestions before merging:
- Use the term "Join" instead of "Group". In Home Assistant, groups typically create new entities, which isn’t the case here. We suggest changing the title to "Join players" and the button label to "Join".
- Instead of dropdown menus, make a list of all media players with checkboxes. This makes it easier to select media players.

As a future enhancement, we could add volume sliders next to each media player in the list.

What are your thoughts on this?

AlexGustafsson commented:
@matthiasdebaat thanks for taking a look! I've played around with using a data table, but I'm not sure it's a great fit. It is used for similar things elsewhere, but I'm not sure how well future features like a volume slider would work. Maybe it would work fine just having it as a separate column. What do you reckon?

![image](https://github.com/user-attachments/assets/c92e80e3-075f-4ba5-9050-54e549bd9225)

matthiasdebaat commented:
I've checked other apps for the best design pattern. I've made a design based on this. What do you think?

As a first iteration we could have a simple toggle switch and a second line showing what the other media players are playing.
![Join media players simple](https://github.com/user-attachments/assets/0324a6cb-05ea-4ae8-87ea-fddd5d3ded3f)

And a second iteration where we would add the volume sliders.
![Join media players](https://github.com/user-attachments/assets/922461b1-f810-44a0-9284-86afa6d513ca)

AlexGustafsson commented:
@matthiasdebaat Thanks! That was really helpful. Demo below.

https://github.com/user-attachments/assets/d69980d9-18f9-47fb-8715-6b75d83d6eab

What I hoped to show with the P2 (classical) radio station at the end was how radio is handled, both when there's a long title and when there's no artist information. But nothing was playing at the time. Here's a still:

![image](https://github.com/user-attachments/assets/ab1df161-c7fd-466a-b786-568b1152333e)

For the developers, I've added a couple of comments to PR stating a couple of concerns I have with the implementation.

As a general note, as a user of the feature, I think it's quite useful already, but it does come with one caveat. We can't really tell which speaker (at least in the case of SONOS) that is the "leader". It seems as if you join speakers that are currently playing, they will stop unless you include its leader. This is a limitation that's there in the SONOS app as well. But it's worth noting. Maybe there's some creative solution, but as a start I guess this will do.

AlexGustafsson commented on src/components/media-player/dialog-join-media-players.ts:
```diff
@@ -0,0 +1,201 @@
+          <span slot="title"
+            >${this.hass.localize("ui.card.media_player.media_players")}</span
+          >
+          <ha-button slot="actionItems" @click=${this._selectAll}>
```
Based on the UI mocks, we'd like the button to be vertically centered with the rest of the items. From what I can tell, the action items container is not vertically centered. This seems to be quite deep in the component hierarchy, so I'm unsure what would be the best solution.

I'm basically after this:

```css
.header-bar {
  align-items: center;
}
```

The current style is to align it to the flex start. Why, I don't know.

silamon replied:
It only seems to be used to display icons. I've left a css suggestion that you can apply to place it vertically.

AlexGustafsson commented on src/components/media-player/ha-media-player-toggle.ts:
```diff
@@ -0,0 +1,89 @@
+        <div class="secondary-text">
+          ${stateObj.state === "playing"
+            ? stateObj.attributes.media_title
+            : this.hass.localize("ui.card.media_player.idle")}
```
We could use the state of the player instead, to handle non-idle cases (they might be unavailable or offline?). But I can't see that the state of media players are used in such a way elsewhere.

Essentially, I'd like to copy the behavior of some "state-visualizing" components that do something like this:

```ts
this.hass.localize(`ui.card.media_player.state.${stateObj.state}`);
```

matthiasdebaat commented:
Awesome! Thanks for your demo, looks great! 
Let's change the button `Join` to `Apply` as it also "unjoins" media players.

AlexGustafsson commented:
@matthiasdebaat thanks, fixed!

@silamon, @bramkragten anything from your side?

silamon commented:
I'll do a review somewhere this week.

silamon commented on src/components/media-player/dialog-join-media-players.ts:
```diff
@@ -0,0 +1,201 @@
+  ) =>
+    Object.values(entities).filter(
+      (entity) =>
+        this._entityId !== undefined &&
+        entity.entity_id !== this._entityId &&
```
```suggestion
        computeDomain(entity.entity_id) === "media_player" &&
        this._entityId !== undefined &&
        entity.entity_id !== this._entityId &&
```

silamon commented on src/components/media-player/ha-media-player-toggle.ts:
```diff
@@ -0,0 +1,97 @@
+      <ha-svg-icon .path=${mdiSpeaker}></ha-svg-icon>
+      <div class="info">
+        <div class="main-text">${computeStateName(stateObj)}</div>
+        <!-- TODO: or state if not playing -->
```
```suggestion
```

It shows already idle if not playing

silamon commented on src/components/media-player/dialog-join-media-players.ts:
```diff
@@ -0,0 +1,201 @@
+    return [
+      haStyleDialog,
+      css`
+        .content {
+          display: flex;
+          flex-direction: column;
+          row-gap: 16px;
+        }
```
```suggestion
        .content {
          display: flex;
          flex-direction: column;
          row-gap: 16px;
        }
        
        ha-dialog-header ha-button {
          margin: 6px;
          display: block;
        }
```

silamon commented on src/components/media-player/dialog-join-media-players.ts:
```diff
@@ -0,0 +1,201 @@
+import "../ha-button";
+import "../ha-dialog-header";
+import "./ha-media-player-toggle";
+import "../../panels/lovelace/components/hui-marquee";
```
```suggestion
```

bramkragten commented on src/components/media-player/dialog-join-media-players.ts:
```diff
@@ -0,0 +1,207 @@
+        computeDomain(entity.entity_id) === "media_player" &&
+        this._entityId !== undefined &&
+        entity.entity_id !== this._entityId &&
+        this.hass.entities[entity.entity_id].platform ===
```
Not every entity has to be in the entity registry, so this could be `undefined` causing an error.

bramkragten commented on src/components/media-player/dialog-join-media-players.ts:
```diff
@@ -0,0 +1,207 @@
+        this._entityId !== undefined &&
+        entity.entity_id !== this._entityId &&
+        this.hass.entities[entity.entity_id].platform ===
+          this.hass.entities[this._entityId].platform &&
```
Not every entity has to be in the entity registry, so this could be undefined causing an error.

bramkragten commented on src/components/media-player/dialog-join-media-players.ts:
```diff
@@ -0,0 +1,207 @@
+        this.hass.entities[entity.entity_id].platform ===
+          this.hass.entities[this._entityId].platform &&
+        supportsFeature(
+          this.hass.states[entity.entity_id],
```
Not every entity in the entity registry has to be in the state machine, we should check that first

AlexGustafsson commented:
@bramkragten Thanks for the look. I've rewritten that quite awful expression to catch that issue and hopefully be a bit clearer.

bramkragten commented on src/components/media-player/dialog-join-media-players.ts:
```diff
@@ -0,0 +1,218 @@
+
+  public closeDialog() {
+    this._entityId = undefined;
+    this._selectedEntities = [];
```
```suggestion
    this._selectedEntities = [];
    this._groupMembers = [];
```

bramkragten commented on src/components/media-player/dialog-join-media-players.ts:
```diff
@@ -0,0 +1,218 @@
+      return [];
+    }
+
+    const currentPlatform = this.hass.entities[this._entityId]?.platform;
```
We should probably make an exception here if `currentPlatform === undefined`

bramkragten commented on src/components/media-player/dialog-join-media-players.ts:
```diff
@@ -0,0 +1,218 @@
+    const currentPlatform = this.hass.entities[this._entityId]?.platform;
+
+    return Object.values(entities).filter((entity) => {
+      const isCurrentEntity = entity.entity_id === this._entityId;
+      const isMediaPlayer = computeDomain(entity.entity_id) === "media_player";
+      const isSamePlatform =
+        this.hass.entities[entity.entity_id]?.platform === currentPlatform;
```
```suggestion
      if(entity.entity_id === this._entityId) {
        return false;
      }
      if(!computeDomain(entity.entity_id) === "media_player") {
        return false;
      }
      if (this.hass.entities[entity.entity_id]?.platform !== currentPlatform) {
        return false;
      }
```

bramkragten commented on src/components/media-player/dialog-join-media-players.ts:
```diff
@@ -0,0 +1,218 @@
+      const isMediaPlayer = computeDomain(entity.entity_id) === "media_player";
+      const isSamePlatform =
+        this.hass.entities[entity.entity_id]?.platform === currentPlatform;
+      const supportsGrouping =
+        this.hass.states[entity.entity_id] &&
+        supportsFeature(
+          this.hass.states[entity.entity_id],
+          MediaPlayerEntityFeature.GROUPING
+        );
+
+      return (
+        !isCurrentEntity && isMediaPlayer && isSamePlatform && supportsGrouping
+      );
```
```suggestion
      if (
        !this.hass.states[entity.entity_id] ||
        !supportsFeature(
          this.hass.states[entity.entity_id],
          MediaPlayerEntityFeature.GROUPING
        )) {
          return false;
        } 
      return true;
```

bramkragten committed:
update

bramkragten committed:
Merge branch 'dev' into media-player-grouping

bramkragten committed:
Update dialog-join-media-players.ts

AlexGustafsson commented:
@bramkragten, @silamon, @matthiasdebaat Thanks a lot for helping me get this in. I really appreciate that you took the time to help out and taught me a thing or two (or a dozen) about the codebase. Don't hesitate to reach out if I can help with issues or improvements related to this feature in the future.

silamon commented:
Thank you for your contribution to this project and your willingness to adopt to the ux designs proposed! One thing that already popped up is that the badge on the join button is not displayed in the more info dialog. But that was not a blocker to get this merged.

bramkragten commented:
> Thank you for your contribution to this project and your willingness to adopt to the ux designs proposed! One thing that already popped up is that the badge on the join button is not displayed in the more info dialog. But that was not a blocker to get this merged.

I added it in this PR ;-)

@AlexGustafsson what @silamon said, thank you very much for your awesome contribution! 🎉
